### PR TITLE
TF to Torch Migrate TimeSeries Feature

### LINKED ISSUE #1323: Migrate time series features to PyTorch

### PR DESCRIPTION
**Pull request by jimthompson5802**
# Code Pull Requests

Closes #1323.

Summary of changes
* Modified `TimeSeriesInputFeature` to work in PyTorch
* Add unit test `test_timeseries_feature.py` 

```
root@4ba4e6a3d53d:/opt/project/sandbox/pytest_area# pytest -v --tb=line /opt/project/tests/integration_tests/test_timeseries_feature.py
=========================================================== test session starts ============================================================
platform linux -- Python 3.7.11, pytest-6.2.4, py-1.10.0, pluggy-0.13.1 -- /usr/local/bin/python
cachedir: .pytest_cache
rootdir: /opt/project, configfile: pytest.ini
plugins: timeout-1.4.2, xdist-2.3.0, cov-2.12.1, forked-1.3.0, pycharm-0.7.0, anyio-3.3.0
collected 6 items

../../tests/integration_tests/test_timeseries_feature.py::test_timeseries_feature[stacked_cnn] PASSED                                [ 16%]
../../tests/integration_tests/test_timeseries_feature.py::test_timeseries_feature[parallel_cnn] PASSED                               [ 33%]
../../tests/integration_tests/test_timeseries_feature.py::test_timeseries_feature[stacked_parallel_cnn] PASSED                       [ 50%]
../../tests/integration_tests/test_timeseries_feature.py::test_timeseries_feature[rnn] PASSED                                        [ 66%]
../../tests/integration_tests/test_timeseries_feature.py::test_timeseries_feature[cnnrnn] PASSED                                     [ 83%]
../../tests/integration_tests/test_timeseries_feature.py::test_timeseries_feature[passthrough] PASSED                                [100%]

============================================================= warnings summary =============================================================
../../../../usr/local/lib/python3.7/site-packages/pyDOE2/doe_factorial.py:16
  /usr/local/lib/python3.7/site-packages/pyDOE2/doe_factorial.py:16: DeprecationWarning: the imp module is deprecated in favour of importlib; see the module's documentation for alternative uses
    import imp

tests/integration_tests/test_timeseries_feature.py::test_timeseries_feature[stacked_cnn]
  /usr/local/lib/python3.7/site-packages/torch/nn/functional.py:652: UserWarning: Named tensors and all their associated APIs are an experimental feature and subject to change. Please do not use them for anything important until they are released as stable. (Triggered internally at  /pytorch/c10/core/TensorImpl.h:1156.)
    return torch.max_pool1d(input, kernel_size, stride, padding, dilation, ceil_mode)

tests/integration_tests/test_timeseries_feature.py::test_timeseries_feature[parallel_cnn]
  /usr/local/lib/python3.7/site-packages/torch/nn/modules/conv.py:295: UserWarning: Using padding='same' with even kernel lengths and odd dilation may require a zero-padded copy of the input be created (Triggered internally at  /pytorch/aten/src/ATen/native/Convolution.cpp:660.)
    self.padding, self.dilation, self.groups)

-- Docs: https://docs.pytest.org/en/stable/warnings.html
====================================================== 6 passed, 3 warnings in 0.30s =======================================================
```

### CODE REVIEW

File: tests/integration_tests/test_timeseries_feature.py [hunk 1]
```diff
@@ -0,0 +1,50 @@
+import pytest
+
+import torch
+
+from ludwig.features.timeseries_feature import TimeseriesInputFeature
+from tests.integration_tests.utils import timeseries_feature
+
+BATCH_SIZE = 2
+SEQ_SIZE = 10
+DEFAULT_FC_SIZE = 256
```
Comment (w4nderlust) [inline tests/integration_tests/test_timeseries_feature.py:10]: We can make this smaller for speed purposes. Maybe 4?

File: tests/integration_tests/test_timeseries_feature.py [hunk 2]
```diff
@@ -0,0 +1,50 @@
...
+                                    dtype=torch.float32)
+
+    # generate feature config
+    timeseries_feature_config = timeseries_feature(
+        encoder=enc_encoder,
+        max_len=SEQ_SIZE,
+        # simulated parameters determined by pre-processing
+        max_sequence_length=SEQ_SIZE,
+        embedding_size=1,
+        should_embed=False
```
Comment (w4nderlust) [inline tests/integration_tests/test_timeseries_feature.py:32]: These 2 paramters shouldn't be needed, as they should be the defaults

File: tests/integration_tests/test_timeseries_feature.py (hunk 1 above)
Comment (jimthompson5802) [reply tests/integration_tests/test_timeseries_feature.py:10]: done.

File: tests/integration_tests/test_timeseries_feature.py (hunk 2 above)
Comment (jimthompson5802) [reply tests/integration_tests/test_timeseries_feature.py:32]: @w4nderlust  
As I understand the code, the two parameters are not "default" per se but are set when the feature's configuration is [updated with results of creating the feature's metadata](https://github.com/jimthompson5802/ludwig/blob/f497e111874fddea10822331762f0cd24d591c3b/ludwig/features/timeseries_feature.py#L178-L187l).   The only [default values set are these](https://github.com/jimthompson5802/ludwig/blob/f497e111874fddea10822331762f0cd24d591c3b/ludwig/features/timeseries_feature.py#L190-L197).

I can implement the comment by moving those two parameters from `update_config_with_metadata()` to `populate_defaults()` method and then calling `TimeSeriesInputFeature.populate_defaults()` in the unit test.  So the unit test code will look something like this.
```
    # generate feature config
    timeseries_feature_config = timeseries_feature(
        encoder=enc_encoder,
        max_len=SEQ_SIZE,
        fc_layers=[{'fc_size': DEFAULT_FC_SIZE}],
        # simulated parameters determined by pre-processing
        max_sequence_length=SEQ_SIZE,
    )

    # instantiate input feature object
    TimeSeriesInputFeature.populate_defaults(timeseries_feature_config)
    timeseries_input_feature = TimeseriesInputFeature(timeseries_feature_config)
```

Does this sound reasonable?

Comment (w4nderlust) [reply tests/integration_tests/test_timeseries_feature.py:33]: You have a good point. The issue though is that those two parameters are parameters for the encoders really, and no other parameter would work. So probably a better idea would be not to set them at all, and just pass them as fixed values when initializing the encoder. Something like
```
self.encoder = encoder_cls(embedding_size=1,  should_embed=False, **other_params)
```

Comment (jimthompson5802) [reply tests/integration_tests/test_timeseries_feature.py]: OK...made requested change and pushed up.  Unit test pass with the changes
```
 ============================= test session starts ==============================
collecting ... collected 6 items

../../tests/integration_tests/test_timeseries_feature.py::test_timeseries_feature[stacked_cnn] 
../../tests/integration_tests/test_timeseries_feature.py::test_timeseries_feature[parallel_cnn] 
../../tests/integration_tests/test_timeseries_feature.py::test_timeseries_feature[stacked_parallel_cnn] 
../../tests/integration_tests/test_timeseries_feature.py::test_timeseries_feature[rnn] 
../../tests/integration_tests/test_timeseries_feature.py::test_timeseries_feature[cnnrnn] 
../../tests/integration_tests/test_timeseries_feature.py::test_timeseries_feature[passthrough] 

======================== 6 passed, 3 warnings in 0.21s =========================
```

**Pull request merged**